File: libs/srs/src/lib/templates.ts
import type { MessageDescriptor } from "@lingui/core";
import { msg } from "@lingui/core/macro";
import { z } from "zod/v4";
import type { Timestamps } from "./db";
import type { UpdateData } from "./utility";

export const templatesMessages: Record<string, MessageDescriptor> = {
  "title.min-length": msg`title.min-length`,
  "title.max-length": msg`title.max-length`,
  "fields.min-length": msg`fields.min-length`,
  "layout.min-length": msg`layout.min-length`,
};

export const TEMPLATE_FIELD_TYPES = ["text"] as const;

export const TEMPLATE_OPERATIONS = ["display", "reveal", "type"] as const;

export type TemplateOperation = typeof TEMPLATE_OPERATIONS[number];

export const templateValidation = z.object({
  id: z.int(),
  title: z
    .string()
    .min(1, "title.min-length")
    .max(255, "title.max-length"),
  content: z.object({
    fields: z.array(z.object({
      id: z.number(),
      title: z.string(),
      type: z.enum(TEMPLATE_FIELD_TYPES),
      isRequired: z.boolean(),
    })).min(1, "fields.min-length"),
    layout: z.array(z.object({
      field: z.number(),
      operation: z.enum(TEMPLATE_OPERATIONS),
    })).min(1, "layout.min-length"),
  }),
});

export type Template = z.infer<typeof templateValidation> & Timestamps & { isLocked: boolean };

export type TemplateFields = Template["content"]["fields"];

export type TemplateField = TemplateFields[number];

export type TemplateLayout = Template["content"]["layout"];

export type TemplateLayoutItem = TemplateLayout[number];

export const DEFAULT_TEMPLATE: InsertTemplateData = {
  title: "Default",
  content: {
    fields: [
      { id: 1, title: "Front", type: "text", isRequired: true },
      { id: 2, title: "Back", type: "text", isRequired: true },
    ],
    layout: [
      { field: 1, operation: "display" },
      { field: 2, operation: "type" },
    ],
  },
};

export const DEFAULT_TEMPLATE_FIELD: TemplateField = {
  id: 0,
  title: "",
  type: "text",
  isRequired: true,
};

/**
 * Gets the title of a template field by its ID
 * @param fields - Array of template fields
 * @param id - The ID of the field to find
 * @returns The title of the field if found, undefined otherwise
 */
export function getTemplateFieldTitleById(
  fields: Template["content"]["fields"],
  id: Template["content"]["fields"][number]["id"],
) {
  return fields.find((x) => x.id === id)?.title;
}

export const insertTemplateSchema = templateValidation.omit({ id: true });

export type InsertTemplateData = z.input<typeof insertTemplateSchema>;

export const updateTemplateSchema = templateValidation.omit({ id: true });

export type UpdateTemplateValues = z.input<typeof updateTemplateSchema>;

export type UpdateTemplateData = UpdateData<Template, "id", UpdateTemplateValues>;

/**
 * Validates that locked template fields have not been modified inappropriately
 * @param original - The original template fields
 * @param updated - The updated template fields
 * @returns Object containing validation result and any errors found
 */
export function validateLockedTemplateFields(original: TemplateField[], updated: TemplateField[]) {
  const errors: string[] = [];

  // check if all fields are present
  const updatedIds = new Set(updated.map(field => field.id));
  const missingIds = original
    .map(field => field.id)
    .filter(id => !updatedIds.has(id));

  if (missingIds.length > 0) {
    errors.push(`Missing fields: ${missingIds.join(", ")}`);
  }

  // check if properties (except 'title') are not changed
  for (const originalField of original) {
    const updatedField = updated.find(f => f.id === originalField.id);
    if (!updatedField) continue;

    const originalKeys = Object.keys(originalField) as (keyof TemplateField)[];
    for (const key of originalKeys) {
      if (key === "title") continue;

      if (updatedField[key] !== originalField[key]) {
        errors.push(`Field (id: ${originalField.id}): property '${key}' changed`);
      }
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
}

export const cloneTemplateSchema = insertTemplateSchema.pick({ title: true }).extend({
  sourceId: templateValidation.shape.id,
});

export type CloneTemplateData = z.infer<typeof cloneTemplateSchema>;

export type DeleteTemplateData = Pick<Template, "id">;

File: apps/native-tauri/src-tauri/src/domain/templates.rs
use serde::{Deserialize, Serialize};

use crate::app::error::{error_codes, AppError};
use crate::app::utility::{serialize_optional_timestamp, serialize_timestamp};

const TITLE_MIN_LENGTH: usize = 1;
const TITLE_MAX_LENGTH: usize = 255;

const FIELD_TYPES: &[&str] = &["text"];
const LAYOUT_OPERATIONS: &[&str] = &["display", "reveal", "type"];

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Template {
    pub id: i64,
    pub title: String,
    pub content: TemplateContent,
    pub is_locked: bool,
    #[serde(serialize_with = "serialize_timestamp")]
    pub created_at: i64,
    #[serde(default, serialize_with = "serialize_optional_timestamp")]
    pub updated_at: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TemplateContent {
    pub fields: Vec<TemplateField>,
    pub layout: Vec<TemplateLayoutItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TemplateField {
    pub id: i64,
    pub title: String,
    #[serde(rename = "type")]
    pub field_type: String,
    pub is_required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TemplateLayoutItem {
    pub field: i64,
    pub operation: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InsertTemplateData {
    pub title: String,
    pub content: TemplateContent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateTemplateValues {
    pub title: String,
    pub content: TemplateContent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateTemplateData {
    pub id: i64,
    pub values: UpdateTemplateValues,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CloneTemplateData {
    pub title: String,
    pub source_id: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteTemplateData {
    pub id: i64,
    pub successor_id: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TemplateDeck {
    pub id: i64,
    pub title: String,
}

impl InsertTemplateData {
    pub fn validate(&self) -> Result<(), AppError> {
        validate_title(&self.title)?;
        validate_template_content(&self.content, None)
    }
}

impl UpdateTemplateValues {
    pub fn validate(&self, original: Option<&TemplateContent>) -> Result<(), AppError> {
        validate_title(&self.title)?;
        validate_template_content(&self.content, original)
    }
}

fn validate_title(title: &str) -> Result<(), AppError> {
    if title.len() < TITLE_MIN_LENGTH {
        return Err(AppError::new(
            error_codes::VALIDATION_COMMON_TITLE_TOO_SHORT,
            Some(format!("Min length: {}", TITLE_MIN_LENGTH)),
        ));
    }

    if title.len() > TITLE_MAX_LENGTH {
        return Err(AppError::new(
            error_codes::VALIDATION_COMMON_TITLE_TOO_LONG,
            Some(format!("Max length: {}", TITLE_MAX_LENGTH)),
        ));
    }

    Ok(())
}

fn validate_template_content(
    content: &TemplateContent,
    original: Option<&TemplateContent>, // Some - template is locked, None - not locked
) -> Result<(), AppError> {
    if content.fields.is_empty() {
        return Err(AppError::new(error_codes::VALIDATION_TEMPLATES_FIELDS_TOO_FEW, None));
    }

    if content.layout.is_empty() {
        return Err(AppError::new(error_codes::VALIDATION_TEMPLATES_LAYOUT_TOO_FEW, None));
    }

    for field in &content.fields {
        if !FIELD_TYPES.contains(&field.field_type.as_str()) {
            return Err(AppError::new(
                error_codes::UNKNOWN,
                Some(format!("Invalid field type: {}", field.field_type)),
            ));
        }
    }

    for item in &content.layout {
        if !LAYOUT_OPERATIONS.contains(&item.operation.as_str()) {
            return Err(AppError::new(
                error_codes::UNKNOWN,
                Some(format!("Invalid layout operation: {}", item.operation)),
            ));
        }
    }

    let field_ids: std::collections::HashSet<i64> = content.fields.iter().map(|f| f.id).collect();
    for item in &content.layout {
        if !field_ids.contains(&item.field) {
            return Err(AppError::new(
                error_codes::UNKNOWN,
                Some(format!("Non-existent field id in layout: {}", item.field)),
            ));
        }
    }

    if let Some(orig) = original {
        validate_locked_template_fields(&orig.fields, &content.fields)?;
    }

    Ok(())
}

fn validate_locked_template_fields(original: &[TemplateField], updated: &[TemplateField]) -> Result<(), AppError> {
    let updated_ids: std::collections::HashSet<i64> = updated.iter().map(|f| f.id).collect();

    // Check if all original fields are present
    for orig_field in original {
        if !updated_ids.contains(&orig_field.id) {
            return Err(AppError::new(
                error_codes::VALIDATION_TEMPLATES_UPDATE_LOCKED,
                Some(format!("Missing field with id: {}", orig_field.id)),
            ));
        }
    }

    // Check if properties (except 'title') are not changed for existing fields
    for orig_field in original {
        if let Some(updated_field) = updated.iter().find(|f| f.id == orig_field.id) {
            // Check type property
            if updated_field.field_type != orig_field.field_type {
                return Err(AppError::new(
                    error_codes::VALIDATION_TEMPLATES_UPDATE_LOCKED,
                    Some(format!("Cannot change field type for field id: {}", orig_field.id)),
                ));
            }
            // Check is_required property
            if updated_field.is_required != orig_field.is_required {
                return Err(AppError::new(
                    error_codes::VALIDATION_TEMPLATES_UPDATE_LOCKED,
                    Some(format!("Cannot change 'isRequired' for field id: {}", orig_field.id)),
                ));
            }
        }
    }

    Ok(())
}
